pwndbg/storage.py

```diff
@@ -0,0 +1,144 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""
+Provides persistent volatile and non-volatile storage of the results of
+a function call.  Volatile storage uses the temporary directory, and is
+generally cleared each reboot, or as the OS dictates.
+
+Cached results are stored per-architecture, i.e. arm and i386 caches
+are independent.  Cached results are also stored per-argument.
+"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import functools
+import inspect
+import os
+import pickle
+import shelve
+import tempfile
+import time
+
+import pwndbg.arch
+import pwndbg.config
+
+nonvolatile = pwndbg.config.Parameter('pwndbg-cache-nonvolatile', os.path.expanduser('~/.pwndbg'), 'Non-volatile cache directory for Pwndbg storage')
+volatile = pwndbg.config.Parameter('pwndbg-cache-volatile', tempfile.gettempdir(), 'Volatile cache directory for Pwndbg storage')
+
+# Ensure there is only one Shelf instance per database, so that
+# we don't get weird things with multiple instances / concurrent writes / etc.
+shelves = {
+    nonvolatile: None,
+    volatile: None
+}
+
+@pwndbg.config.Trigger([volatile, nonvolatile])
+def update():
+    global shelves
+
+    for db in (volatile, nonvolatile):
+        database_path = os.path.join(str(db), 'storage')
+
+        # Ensure the path exists, then open the database
+        if not os.path.isdir(database_path):
+            os.makedirs(database_path)
+
+        shelves[db] = shelve.open(database_path,
+                                  writeback=True,
+                                  protocol=pickle.HIGHEST_PROTOCOL)
+
+# Perform manual update to initialize the values
+update()
+
+class Cached(object):
+    """Function decorator that persistently caches the results of the function it decorates."""
+
+    def __init__(self, function, type):
+        self.function = function
+        self.function_name = function.__module__ + '.' + function.__name__
+        self.type = Type
```

Comment:
`Type` is undefined. Also, `type` shadows the builtin, so if we do care, maybe `type_`